adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,551 @@
+CUDA devices available in the system.
+[[sec:cuda:introduction]]
+== Introduction
+
+[[sec:cuda:mapping_of_sycl_programming_model]]
+== Mapping of SYCL programming model
+
+This section gives a general overview of how the SYCL programming model maps to
+CUDA. These two programming models are pretty similar in essence however they do
+have a few differences in terminology and architecture.
+
+[[sub:cuda:platform_model]]
+=== Platform Model
+
+A SYCL device maps to a single CUDA device.  As CUDA does not split into
+separate platforms there is no 'platform' concept in CUDA corresponding to the
+SYCL platform. Instead, a SYCL platform maps to a collection of CUDA devices
+represented by [code]#std::vector<CUdevice>#.
+
+A SYCL <<context>> simply maps to one, or multiple CUDA contexts. Indeed while
+a CUDA context is tied to a single device, this is not the case for a SYCL
+<<context>> and the CUDA backend implementation may use multiple CUDA contexts
+to emulate a SYCL <<context>> containing multiple devices.
+
+In CUDA, contexts and devices may need to be set as active on a thread for the
+CUDA Driver or CUDA Runtime APIs to use. Therefore the SYCL API entry points
+may change the active context or device whenever required for proper execution
+of SYCL operations, and no guarantee is provided that any prior active context
+or active device will be restored by the SYCL API entry points before returning
+to the user application.
+
+A SYCL <<queue>> simply maps to one or multiple CUDA streams. Indeed while a
+CUDA stream is in-order, a SYCL <<queue>> isn't, so a CUDA backend implementation
+may use multiple CUDA streams to implement an out of order SYCL <<queue>>.
+
+[[sub:cuda:memory_model]]
+=== Memory model
+
+==== Accessing memory on a different GPU
+
+Devices belonging to the same context must be able to access (directly or indirectly) each other's global memory. This is done in one of the following ways:
+
+- Device directly accesses memory on another device (peer-to-peer memory access).
+- CUDA-runtime-managed memory is used. CUDA runtime copies the data from one device to another.
+- Peer copy (one of [code]#CuMemcpyPeerAsync# or [code]#CuMemcpy3DPeerAsync#) is used to directly copy the data from one device to another.
+- The data is copied from one device to the host, and then from host to another device.
+
+==== Shared USM memory advices
+
+Values for the [code]#advice# parameter of [code]#sycl::queue::mem_advise# and [code]#sycl::handler::mem_advise# and their mapping to CUDA equivalent are defined in table <<table.cuda.memmodel.advices>>.
+
+[[table.cuda.memmodel.advices]]
+.Valid shared USM advices and their equivalents in CUDA
+[width="100%",options="header",cols="40%,30%,30%"]
+|====
+| SYCL shared USM advice | CUDA managed memory advice | processor, the advice is set for
+| [code]#sycl::cuda::advice::cuda_mem_advise_set_read_mostly# | [code]#cudaMemAdviseSetReadMostly# | device associated with the queue/handler
```

Comment:
@AerialMantis FYI these `advise` names seem to be different than those used in DPC++: https://github.com/t4c1/llvm-test-suite/blob/1a5a4e9d72cccf290d88f2498c08314d8bfb271e/SYCL/USM/memadvise_cuda.cpp#L39

Do we need to update one or other of the documentation / dpc++ impl?

Reply:
Good catch, these are different, but I think we may want to update DPC++ to match what we have in here. I have also updated the definitions here to follow the correct extension convention.